TF2 environment settings (#5230)
RasaHQ/rasa: docs/api/tensorflow_usage.rst

:desc: Find out how to configure your environment for efficient usage of TensorFlow inside Rasa OSS

.. _tensorflow_usage:

Setting up the TensorFlow Runtime
=================================

TensorFlow allows setting the runtime environment via
`TF Config submodule <https://www.tensorflow.org/api_docs/python/tf/config>`_. Rasa OSS supports a smaller subset of these
configuration options and makes appropriate calls to the ``tf.config`` submodule.

> This smaller subset comprises of configurations that developers frequently use with Rasa OSS.

Reviewer: ```suggestion
This smaller subset comprises of configurations that developers frequently use with Rasa Open Source.
```

All configuration options are specified using environment variables as shown in subsequent sections.

Optimizing CPU Performance
--------------------------

> Parallelizing one operation

Reviewer 2: ```suggestion
Parallelizing One Operation
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^

Set ``TF_INTRA_OP_PARALLELISM_THREADS`` as an environment variable to specify maximum number of threads that can be used
to parallelize the execution of one operation. If left unspecified, this value defaults to ``0`` which means TensorFlow should
pick an appropriate value depending on the system configuration.

> 

Reviewer 2: optional
```suggestion
```

> Parallelizing multiple operations

Reviewer 2: ```suggestion
Parallelizing Multiple Operations
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Set ``TF_INTER_OP_PARALLELISM_THREADS`` as an environment variable to specify maximum number of threads that can be used
to parallelize the execution of multiple **non-blocking** operations. If left unspecified, this value defaults to ``0``
which means TensorFlow should pick an appropriate value depending on the system configuration.

> 

Reviewer 2: optional
```suggestion
```

Optimizing GPU Performance
--------------------------

> Limiting GPU memory growth

Reviewer 2: ```suggestion
Limiting GPU Memory Growth
```

^^^^^^^^^^^^^^^^^^^^^^^^^^

TensorFlow by default blocks all the available GPU memory for the running process. This can be limiting if you are running
multiple TensorFlow processes and want to distribute memory across them. To prevent this,

> set an environment variable ``TF_FORCE_GPU_ALLOW_GROWTH`` to ``True``.

Reviewer 2: `TF_FORCE_GPU_ALLOW_GROWTH` means the gpu is forced to grow to use more memory than available or how can is understand that? Or should this be set to `False`?

Author: @wochinge This is an environment variable that Tensorflow provides and I agree that unfortunately it isn't coined very well. So, if the value is `False`(default value), then the process blocks the complete GPU memory available. If it is `True` it would allocate memory as required and would **grow** only if required. In no way, the process that use more memory than the GPU provides it.

Reviewer 2: Wild 😄

> 

Reviewer 2: optional
```suggestion
```

> Restricting absolute GPU memory available

Reviewer 2: ```suggestion
Restricting Absolute GPU Memory Available
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Often, a developer wants to limit the absolute amount of GPU memory that can be used by a process.

For example, you may have two visible GPUs(``GPU:0`` and ``GPU:1``) and you want to allocate 1024 MB from the first GPU
and 2048 MB from the second GPU.
You can do so by setting an environment variable as ``TF_GPU_MEMORY_ALLOC="0:1024, 1:2048"``.

> Another scenario can be where you have access to 2 GPUs(``GPU:0`` and ``GPU:1``) but you would like to use only second

Reviewer: ```suggestion
Another scenario can be where you have access to 2 GPUs(``GPU:0`` and ``GPU:1``) but you would like to use only second GPU
```

> GPU for the Rasa OSS process.

Reviewer: ```suggestion
GPU for the Rasa Open Source process.
```

Reviewer: it sounds stupid here

Reviewer: `for the rasa process` since it is the name of actual library?

Reviewer: ```suggestion

```

> ``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory availble from GPU 1.

Reviewer 2: ```suggestion
``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory available from GPU 1.
```

> ``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory from GPU 1 available for the Rasa process

Reviewer: ```suggestion
``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory from GPU 1 available for the Rasa Open Source process
```

Reviewer: ```suggestion
``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory from GPU 1 available
```

:desc: Find out how to configure your environment for efficient usage of tensorflow inside Rasa

.. _tensorflow_usage:

Setting up Tensorflow Runtime
=============================

Tensorflow allows setting the runtime environment via

> `TF Config submodule <https://www.tensorflow.org/api_docs/python/tf/config>`_. Rasa supports a smaller subset of these

Author: Changed to Rasa OSS everywhere

> `TF Config submodule <https://www.tensorflow.org/api_docs/python/tf/config>`_. Rasa supports a smaller subset of these

Author: Reverted to Rasa Open Source 😄

configuration options and makes appropriate calls to the `tf.config` submodule.
This smaller subset comprises of configurations that developers frequently use with Rasa.
All configuration options are specified using environment variables as shown in subsequent sections.

Optimize CPU Performance
------------------------

> Parallelize one operation

Author: A matrix multiplication can be parallelized on multiple threads. It can be a single operation but a non-atomic one. :)
